File: src/utils/nyaa.ts
import { parse as parseHtml } from "node-html-parser";
import { request } from "undici";

export type NyaaSearchItemStatus = "success" | "default" | "danger";

export interface NyaaSearchItem {
    id: string;
    status: NyaaSearchItemStatus;
    category: string;
    title: string;
    url: string;
    torrent: string;
    magnet: string;
    size: string;
    date: string;
    seeders: number;
    leechers: number;
    downloads: number;
}

export interface NyaaSearchResult {
    url: string;
    terms: string;
    items: NyaaSearchItem[];
}

export interface NyaaMagnetResult {
    id: string;
    title: string;
    url: string;
    magnet: string;
}

export class Nyaa {
    static baseUrl = "https://nyaa.si";
    static faviconUrl = "https://nyaa.si/static/favicon.png";

    static async search(terms: string) {
        const url = `${this.baseUrl}/?q=${encodeURIComponent(
            terms,
        )}&s=seeders&o=desc`;
        const resp = await request(url);
        const html = await resp.body.text();
        const document = parseHtml(html);
        const rows = document.querySelectorAll(".torrent-list tbody tr");
        const items = rows.slice(0, 5).map((x) => {
            const [
                category,
                name,
                links,
                size,
                date,
                seeders,
                leechers,
                downloads,
            ] = x.querySelectorAll("td");
            const nameAnchor = name!.querySelectorAll("a").at(-1);
            const linkAnchors = links!.querySelectorAll("a");
            const url = this.baseUrl + nameAnchor!.getAttribute("href")!;
            const result: NyaaSearchItem = {
                id: this.parseIdFromViewUrl(url),
                status: x.classList.value[0] as NyaaSearchItemStatus,
                category:
                    category?.querySelector("a")?.getAttribute("title") ?? "?",
                title: nameAnchor!.getAttribute("title")!,
                url: url,
                torrent: this.baseUrl + linkAnchors[0]!.getAttribute("href")!,
                magnet: linkAnchors[1]!.getAttribute("href")!,
                size: size?.textContent ?? "?",
                date: date?.textContent ?? "?",
                seeders: parseInt(seeders?.textContent ?? "") || 0,
                leechers: parseInt(leechers?.textContent ?? "") || 0,
                downloads: parseInt(downloads?.textContent ?? "") || 0,
            };
            return result;
        });
        const result: NyaaSearchResult = {
            url,
            terms,
            items,
        };
        return result;
    }

    static async magnet(id: string) {
        const url = `${this.baseUrl}/view/${id}`;
        const resp = await request(url);
        const html = await resp.body.text();
        const document = parseHtml(html);
        const panel = document.querySelector(".container .panel")!;
        const title = panel.querySelector(".panel-title")!.textContent.trim();
        const magnet = panel
            .querySelector(".panel-footer .card-footer-item")!
            .getAttribute("href")!;
        const result: NyaaMagnetResult = {
            id,
            title,
            url,
            magnet,
        };
        return result;
    }

    static parseIdFromViewUrl(url: string) {
        return url.match(/(\d+)$/)![1]!;
    }
}

File: utils/nyaa.go
package utils

import (
	"fmt"
	"net/http"
	"net/url"
	"regexp"
	"strconv"

	"github.com/PuerkitoBio/goquery"
)

type NyaaSearchItem struct {
	Id        string
	Category  string
	Title     string
	Url       string
	Torrent   string
	Magnet    string
	Size      string
	Date      string
	Seeders   int
	Leechers  int
	Downloads int
}

type NyaaSearchResult struct {
	Url   string
	Terms string
	Items []NyaaSearchItem
}

type NyaaMagnetResult struct {
	Id     string
	Title  string
	Url    string
	Magnet string
}

const NyaaBaseUrl = "https://nyaa.si"
const NyaaFaviconUrl = "https://nyaa.si/static/favicon.png"

func NyaaSearch(terms string) (*NyaaSearchResult, error) {
	url := fmt.Sprintf("%s/?q=%s&s=seeders&o=desc", NyaaBaseUrl, url.QueryEscape(terms))
	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, err
	}
	items := []NyaaSearchItem{}
	doc.Find(".torrent-list tbody tr").Each(func(i int, x *goquery.Selection) {
		rows := x.Find("td")
		name := rows.Eq(1).Find("a").Eq(-1)
		links := rows.Eq(2).Find("a")
		url := NyaaBaseUrl + name.AttrOr("href", "")
		items = append(items, NyaaSearchItem{
			Id:        parseIdFromViewUrl(url),
			Category:  rows.Eq(0).Find("a").AttrOr("title", ""),
			Title:     name.Text(),
			Url:       url,
			Torrent:   NyaaBaseUrl + links.Eq(0).AttrOr("href", ""),
			Magnet:    links.Eq(-1).AttrOr("href", ""),
			Size:      rows.Eq(3).Text(),
			Date:      rows.Eq(4).Text(),
			Seeders:   atoiOrZero(rows.Eq(5).Text()),
			Leechers:  atoiOrZero(rows.Eq(6).Text()),
			Downloads: atoiOrZero(rows.Eq(7).Text()),
		})
	})
	return &NyaaSearchResult{
		Terms: terms,
		Url:   url,
		Items: items,
	}, nil
}

func NyaaMagnet(id string) (*NyaaMagnetResult, error) {
	url := fmt.Sprintf("%s/view/%s", NyaaBaseUrl, id)
	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, err
	}
	panel := doc.Find(".container .panel").Eq(0)
	return &NyaaMagnetResult{
		Id:     id,
		Title:  panel.Find(".panel-title").Eq(0).Text(),
		Url:    url,
		Magnet: panel.Find(".panel-footer .card-footer-item").Eq(0).AttrOr("href", ""),
	}, nil
}

var idRegex = regexp.MustCompile(`\d+$`)

func parseIdFromViewUrl(url string) string {
	return idRegex.FindString(url)
}

func atoiOrZero(value string) int {
	if parsed, err := strconv.Atoi(value); err == nil {
		return parsed
	}
	return 0
}
